Fix #2405 and #2414 related to PDF hyperlinks to numbered code listings

jfbu opened the pull request:
This needs review because I am not sure whether my modification to LaTeX writer `latex.py` does not break something.

jfbu committed:
Fix #2405: numref link in PDF jumps to the wrong location

The hypertarget insertion by the LaTeX writer for literal blocks is
moved from depart_container to visit_caption.

	modified:   sphinx/writers/latex.py

jfbu committed:
Fix broken numeric reference in PDF hyperlinks to "Listing " code-blocks

The `\label` is now inserted into the caption on top of the Verbatim,
and `\ref` will work as expected. No need for `\phantomsection`.

	modified:   sphinx/writers/latex.py
	modified:   tests/test_directive_code.py

jfbu committed:
Remove unneeded test "if 'ids':" from previous commit.

	modified:   sphinx/writers/latex.py

jfbu committed:
Merge branch 'stable' into jumptorightplace-onstable

jfbu committed:
Fix two issues in latex.py reported by Travis CI

	modified:   sphinx/writers/latex.py

jfbu committed:
Fix line too long issue reported by Travis CI

	modified:   sphinx/writers/latex.py

tk0miya commented:
It works good. But, unfortunately, I found the another bug.

Sphinx also allows to refer uncaptioned literal block::

```
:ref:`literal block <label>`

.. literalinclude:: filename.ext
   :name: label
```

But current release and this fix do not support such situation. We have to put label even if there is no captions. 

I think it would be better if the labels will be placed inside `verbatim` environment.

Thanks,

jfbu commented:
@tk0miya thanks for looking.

I will look at un-captioned situation. I indeed feared such a thing was possible in Sphinx, but had not used it enough so far to be sure offhand. Regarding placing label inside verbatim, but then the PDF viewer will jump to below caption, which is bad because user will not see anymore `Listing 2.3` (or whatever) say. The correct place for label in case of caption is within caption as in my PR.

In case of no caption the `\phantomsection\label{...}` way can be used then. If the `\phantomsection\label{..}` is right before the  \begin{Verbatim} after the commands used for ensuring sufficient space on the page, the PDF viewer will jump to correct place.

jfbu committed:
Merge branch 'stable' into jumptorightplace-onstable

jfbu commented:
oh sorry I see it works as `:ref:`See There <bar>`` where the `<` and `>` are explicitely there. And indeed it works for html output, not for pdf.

tk0miya commented:
Indeed, it might be better to change the place of labels where it has caption or not.

I'll take a look this after other critical bugs.

tk0miya commented on sphinx/writers/latex.py:
```diff
@@ -1991,14 +1992,15 @@ def visit_container(self, node):
             self.body.append('\n')
-            self.context.append(ids + '\n')
+            # context.pop will be done in visit_caption
+            self.context.append(ids)
```
Pushing to and popping from context should be symmetrical.
It would be better to move these `ids` generation to `visit_caption`.

jfbu replied:
@tk0miya New version does not use context, but interacts via a LaTeX macro which is checked for by Verbatim. In this way I can make functional links of the `:ref:`Go there <label>`` type, with no caption. I discovered that "visit_container" is not gone through if no caption, hence moved to "visit_literal_block" the handling in case of no caption.

Here is my test file:

``` rst
TEST
====

FIRST
-----

The :numref:`toolbox.code_sdo1` code demonstrates how to solve this problem.
See also :ref:`Here <foo1>`, :numref:`This is foo`,
:numref:`foo2`, and also :ref:`There <foo3>` and :ref:`foo4`.

.. literalinclude:: sdo1.m
   :name: toolbox.code_sdo1
   :caption: Code implementing problem


SECOND
------

This is standard code-block (no name no caption)::

>>> x=y

This is a literal include with name ``foo1`` but without caption.

.. literalinclude:: sdo1.m
   :name: foo1

..

   This should be indented. A literal include with caption but no name.

   .. literalinclude:: sdo1.m
      :caption: This is foo

   ..

      This one has both caption and name.

      .. literalinclude:: sdo1.m
         :name: foo2
         :caption: This is foo2

THIRD
-----

Here a code-block directive with name and without caption.

.. code-block:: python
   :name: foo3

   >>> Hello=['name but no caption']

Here a code-block directive with caption and without name.

.. code-block:: python
   :caption: foo4

   >>> Hello=['caption but no name']

Here is a code-block directive with neither caption nor name.

.. code-block:: python

   >>> Hello=['no caption no name']
```

jfbu committed:
Enable pdf hyperlinks to named literal-blocks without caption.

This is tango between latex.py and sphinx.sty to handle PDF hyperlinks
to literal-blocks (both code-block and literalinclude) either with name
or caption (then numref can be used) or both.

	modified:   sphinx/texinputs/sphinx.sty
	modified:   sphinx/writers/latex.py

This test file is now back to original.

	modified:   tests/test_directive_code.py

jfbu committed:
Fix comment

	modified:   sphinx/writers/latex.py

jfbu commented:
I have commented the changes in a Note to commit https://github.com/sphinx-doc/sphinx/pull/2415/commits/261ff1a9d2f1794f72c275718f31555089fd7e99. Sorry I meant to say my explanations were added to your comment above. I got confused by interface.

jfbu commented:
Regarding 

``` latex
\newcommand{\literalblockwithoutcaptionneedspace}{1.5\baselineskip}
```

at bottom of sphinx.sty, I chose the `1.5\baselineskip` a bit at random. The idea is to guarantee via use of `\needspace` that non-captioned literal-blocks at bottom of a page, if moved to next page for lack of space for at least one line, will not be separated from the possible `\phantomsection\label{foo}` hence the PDF  hyperlink will not jump to bottom of previous page.

One could also in `sphinx.sty` inside definition of Verbatim environment do:

``` latex
  \ifx\SphinxVerbatimTitle\empty
      % no caption. 
      % if there is a label insert hypertarget. Require space if at bottom of page
      \ifx\SphinxLiteralBlockLabel\empty\else
         \needspace{\literalblockwithoutcaptionneedspace}%
          \phantomsection\SphinxLiteralBlockLabel
      \fi
  \fi
```

rather than

``` latex
  \ifx\SphinxVerbatimTitle\empty
      % no caption. Require space if at bottom of page
      \needspace{\literalblockwithoutcaptionneedspace}%
      % if there is a label insert hypertarget.
      \ifx\SphinxLiteralBlockLabel\empty\else
          \phantomsection\SphinxLiteralBlockLabel
      \fi
  \fi
```

This would avoid modifying pdf output for existing documents for non captioned literal-blocks having currently only one line at bottom of page. Current PR has possibly the effect to move that lonely line literal block to next page, because together with  the frame it might still fit in less than `1.5\baselineskip` but my code above requires `1.5\baselineskip`. Or, reduce `1.5\baselineskip` to perhaps `1.2\baselineskip`.

tk0miya commented on sphinx/writers/latex.py:
```diff
@@ -1833,6 +1840,8 @@ def warner(msg):
             self.body.append('\n' + hlcode + '\\end{%sVerbatim}\n' %
                              (self.table and 'Original' or ''))
+            if ids:
+                self.body.append('\\let\\SphinxLiteralBlockLabel\empty\n')
```
`\SphinxVerbatimTitle` is reset on `\endVerbatim`. Is there any reason to reset `\SphinxLiteralBlockLabel` explicitly?

jfbu replied:
The definition of `\SphinxLiteralBlockLabel` is done before `Verbatim` environment is entered in the produced LaTeX file. It is reset after the `Verbatim` environment has ended. This resetting is needed because next literal block will check the status of `\SphinxLiteralBlockLabel` in case no title was explicitly set. Thus if we did not reset to `\empty` that macro a second target with the same label could be created by another literal block with no caption.

jfbu replied:
It is true that `\global\let\SphinxVerbatimTitle\empty` is done from inside `\endVerbatim` (as configured in `sphinx.sty` package). On the other hand the `\let\SphinxLiteralBlockLabel\empty` is inserted by the LaTeX writer after `\end{Verbatim}`.  Thus the two are treated a bit differently. Actually the better way is for both to be reset after the `\end{Verbatim}`. This way no `\global` is needed and possibility of nesting exist (which don't truly exist because after all we can't really nest Verbatim things...). The reason for handling the `\SphinxVerbatimTitle` the way it is is purely because at the time I did the earlier commit which brought it I had not dared looked at `latex.py` but was only working with modifying `sphinx.sty`. I should probably now make an additional change to treat it the same way the `\SphinxLiteralBlockLabel` is handled in the commits of this PR.

tk0miya replied:
+1 for additional change.
I prefer explicit resetting rather than implicit one.

tk0miya commented:
New code does not need any testcases?

jfbu commented:
At some point in this PR, I modified some test because a `\label` was inserted explicitely. Later on with `\SphinxLiteralBlockLabel` thing, I could reset the test to its original version, because the `\label` was handled differently and glued to the caption only if one existed from inside the `Verbatim` code as defined by `sphinx.sty` . A new test could be added to check that a correct `\def\SphinxLiteralBlockLabel{\label{...}}` appears in the produced `.tex` file. Should I do this ?

tk0miya commented:
I think testcases for labels are necessary for us. We have to check both cases; captions exist or not exist.
Could you add the testcases?

jfbu committed:
Make reset of \SphinxVerbatimTitle to \empty explicit in produced LaTeX files

	modified:   sphinx/texinputs/sphinx.sty
	modified:   sphinx/writers/latex.py

jfbu committed:
Remove unneeded \bgroup/\egroup in legacy sphinx.sty definition of Verbatim

Indeed, sphinx.sty redefines _environment_ Verbatim, and LaTeX's
environments always are scope-limiting for local assignments.

	modified:   sphinx/texinputs/sphinx.sty

jfbu committed:
Added Tests for latex(pdf) links to captioned or named literal blocks

As conf.py has now numfig = True, needed to modify accordingly test for
links in html output.

This topic branch fixes #2405, #2414, and also makes for latex(pdf)
functional hyperlinks to named references, but only for literal blocks
(code-block or literalinclude).

	modified:   tests/roots/test-directive-code/caption.rst
	modified:   tests/roots/test-directive-code/conf.py
	new file:   tests/roots/test-directive-code/namedblocks.rst
	modified:   tests/test_directive_code.py

tk0miya commented:
I merged this now. Thank you for great work!
